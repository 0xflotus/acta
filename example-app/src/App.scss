#main {
  padding-bottom: 100px;

  > header {
    > h1 {
      font-size: 32px;
      font-weight: bold;
      line-height: 1;
      width: 300px;
      padding: 20px;
    }
  }

  > .todoLists {
    border-bottom: 1px solid #000;
    display: flex;
    flex-direction: column;

    @media (min-width: 800px) {
      flex-direction: row;
    }

    > div {
      border-top: 1px solid #000;
      flex-grow: 0;
      flex-shrink: 0;

      @media (min-width: 800px) {
        width: 50%;

        &:first-of-type {
          border-right: 1px solid #000;
        }
      }

      > h2 {
        border-bottom: 1px solid #000;
        font-size: 24px;
        font-weight: bold;
        padding: 20px;
      }

      > p {
        color: orange;
        padding: 20px;
      }

      > div {
        display: flex;
        border-bottom: 1px solid #000;
        align-items: stretch;
        justify-content: space-between;

        &:last-of-type {
          border-bottom: none;
        }

        > div {
          padding: 10px;

          > span {
            display: block;

            &:first-of-type {
              font-size: 14px;
            }

            &:last-of-type {
              color: #666;
              font-size: 9px;
            }
          }
        }

        > button {
          border: none;
          padding: 10px 20px;
          font-weight: bold;
          text-transform: uppercase;
          font-size: 10px;
          background: #eee;
          letter-spacing: 1px;
          transition: filter 200ms ease-in-out;

          &:hover {
            filter: brightness(0.75);
          }
        }
      }
    }
  }

  > form {
    padding: 20px;
    border-bottom: 1px solid #000;
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    background-color: #eee;

    > h3 {
      width: 100%;
      font-weight: bold;
      margin-bottom: 5px;
    }

    > input[type='text'] {
      display: block;
      width: 100%;
      border: 1px solid #000;
      padding: 10px;
      font-size: 14px;
      margin-bottom: 10px;
    }

    > p {
      font-size: 12px;
      display: flex;
      align-items: center;
      margin-bottom: 10px;

      > input[type='checkbox'] {
        display: block;
        margin-left: 5px;
        appearance: none;
        width: 15px;
        height: 15px;
        background-color: #fff;
        border: 1px solid #000;
        position: relative;
        cursor: pointer;

        &:checked {
          &:after {
            content: '';
            display: block;
            position: absolute;
            width: 11px;
            height: 11px;
            top: 1px;
            left: 1px;
            background-color: #000;
          }
        }
      }
    }

    > input[type='submit'] {
      border: none;
      background: #000;
      padding: 10px 20px;
      font-weight: bold;
      text-transform: uppercase;
      font-size: 10px;
      letter-spacing: 1px;
      cursor: pointer;
      color: #fff;
    }
  }

  > .messagesContainer {
    position: fixed;
    bottom: 0;
    z-index: 9;

    &.functional {
      left: 0;
    }

    &.class {
      right: 0;
    }

    > div {
      margin: 10px;
      padding: 10px;
      background-color: #eee;
      border: 1px solid #000;
      border-left: 5px solid #000;
    }
  }
}
