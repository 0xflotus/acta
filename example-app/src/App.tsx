import React from 'react';

import { FunctionalTodosList } from './components/FunctionalTodosList';
import { ClassTodoList } from './components/ClassTodoList';
import { AddTodoForm } from './components/AddTodoForm';

export const App = () => {
  return (
    <div className='App'>
      <FunctionalTodosList />
      <ClassTodoList />

      <AddTodoForm />
    </div>
  );
};
