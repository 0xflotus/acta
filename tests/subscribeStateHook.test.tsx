// @ts-nocheck
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import App, { setHookValueInActa } from './testApp';

describe('Acta useActaState hook test', () => {
  /**
   * Feature
   */
  test('When a state change, the value in the the component is updated.', () => {
    let app: ReactTestRenderer;
    renderer.act(() => {
      app = renderer.create(<App />);
    });

    // Render the app with an empty elements array
    // @ts-ignore
    let tree = app.toJSON();

    // Get the target paragraph
    let functionnalComponentRoot = tree?.children?.find(
      // @ts-ignore
      (child) => child.props.id === 'functionnalComponentRoot'
    );
    // @ts-ignore
    let valueFromStateParagraph = functionnalComponentRoot.children.find(
      // @ts-ignore
      (child) => child.props.id === 'valueFromState'
    );

    // Check that we have the initial value
    expect(valueFromStateParagraph.children[0]).toBe('Not set yet');
    expect(tree).toMatchSnapshot();

    // Set the value in Acta state
    // It should trigger the callback
    renderer.act(() => {
      setHookValueInActa();
    });

    // Render the app with an elements array containing strings now
    tree = app.toJSON();

    // Get the target paragraph
    functionnalComponentRoot = tree?.children?.find(
      // @ts-ignore
      (child) => child.props.id === 'functionnalComponentRoot'
    );
    // @ts-ignore
    valueFromStateParagraph = functionnalComponentRoot.children.find(
      // @ts-ignore
      (child) => child.props.id === 'valueFromState'
    );

    // Check that we have the final value
    expect(valueFromStateParagraph.children[0]).toBe(
      'Value from updated state'
    );
    expect(tree).toMatchSnapshot();
  });

  //   test('When a component has already subscribed to a state, a new subscritpion return false', () => {
  //     const actaSubs = Acta.states[ACTA_STATE_ELEMENTS_LIST].subscribtions;
  //     const alreadySubscribedContext = actaSubs[Object.keys(actaSubs)[0]]
  //       .context as IComponentWithID;

  //     expect(
  //       Acta.subscribeState(
  //         ACTA_STATE_ELEMENTS_LIST,
  //         () => true,
  //         alreadySubscribedContext
  //       )
  //     ).toBeUndefined();
  //   });

  //   test('When a component unmounts, it should not be in the acta subs anymore', () => {
  //     const actaSubs = Acta.states[ACTA_STATE_ELEMENTS_LIST].subscribtions;
  //     const alreadySubscribedContext = actaSubs[Object.keys(actaSubs)[0]]
  //       .context as IComponentWithID;

  //     // We should have one sub
  //     expect(Object.keys(actaSubs).length).toBe(1);

  //     if (
  //       alreadySubscribedContext &&
  //       alreadySubscribedContext.componentWillUnmount
  //     ) {
  //       alreadySubscribedContext?.componentWillUnmount();
  //     }

  //     // We should not has subs anymore
  //     expect(Object.keys(actaSubs).length).toBe(0);
  //   });

  //   /**
  //    * Errors management
  //    */
  //   test('When the stateKey param is not a valid string, should throw an error', () => {
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState(null, () => true, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('', () => true, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState([], () => true, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState({}, () => true, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       Acta.subscribeState(
  //         // @ts-ignore
  //         () => true,
  //         // @ts-ignore
  //         () => true,
  //         // @ts-ignore
  //         {} as IComponentWithID
  //       );
  //     }).toThrowError(paramsErrorMessage);
  //   });

  //   test('When the callback param is not a valid function, should throw an error', () => {
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', null, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', undefined, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', [], {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', {}, {} as IComponentWithID);
  //     }).toThrowError(paramsErrorMessage);
  //   });

  //   test('When the context is not param is not an object, should throw an error', () => {
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', () => true);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', () => true, '');
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', () => true, []);
  //     }).toThrowError(paramsErrorMessage);
  //     expect(() => {
  //       // @ts-ignore
  //       Acta.subscribeState('testKey', () => true, null);
  //     }).toThrowError(paramsErrorMessage);
  //   });

  //   test('When the initial value is an object containing a circular object, should return false', () => {
  //     const circularObject: { item?: object } = {};
  //     circularObject.item = circularObject;

  //     expect(
  //       Acta.subscribeState(
  //         'testKey',
  //         () => true,
  //         {} as IComponentWithID,
  //         circularObject
  //       )
  //     ).toBe(undefined);
  //   });
});
