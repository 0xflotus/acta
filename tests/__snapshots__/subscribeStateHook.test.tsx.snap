// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Acta useActaState hook test When a state change, the value in the the component is updated. 1`] = `
<div>
  <h1>
    Elements
  </h1>
  <ul
    id="listContainer"
  />
  <p
    id="simpleValue"
  >
    Not Set
  </p>
  <p
    id="stateWithInitialValue"
  />
  <div
    id="functionnalComponentRoot"
  >
    <p
      id="valueFromState"
    >
      Not set yet
    </p>
    <p
      id="valueFromEvent"
    >
      Not set yet
    </p>
  </div>
  <button
    id="unmountFunctionalComponent"
    onClick={[Function]}
  >
    unmount functionnal component
  </button>
</div>
`;

exports[`Acta useActaState hook test When a state change, the value in the the component is updated. 2`] = `
<div>
  <h1>
    Elements
  </h1>
  <ul
    id="listContainer"
  />
  <p
    id="simpleValue"
  >
    Not Set
  </p>
  <p
    id="stateWithInitialValue"
  />
  <div
    id="functionnalComponentRoot"
  >
    <p
      id="valueFromState"
    >
      Value from updated state
    </p>
    <p
      id="valueFromEvent"
    >
      Not set yet
    </p>
  </div>
  <button
    id="unmountFunctionalComponent"
    onClick={[Function]}
  >
    unmount functionnal component
  </button>
</div>
`;
